[core][gpu-objects] Add initial docs (#55981)
ray-project/ray: doc/source/ray-core/gpu-objects.rst

.. _gpu-objects:

GPU Objects
===========

Ray objects are normally stored in Ray's CPU-based object store and are copied to GPU memory as needed by the user.
This can lead to unnecessary and expensive data transfers.
For example, passing a CUDA ``torch.Tensor`` from one Ray task to another would require a copy from GPU to CPU memory, then back again to GPU memory.

> GPU objects are a new feature that allows Ray to store and pass objects directly in GPU memory, enabling zero-copy or near-zero-copy data transfers.

Reviewer: "zero-copy data transfers" seems a little misleading/confusing -- there's still one copy from GPU<>GPU

maybe be a bit more direct and say "using optimized transports that avoid unnecessary copies to CPU memory"

Reviewer: actually, I'd suggest just deleting this sentence. it's redundant with the below, which is more precise

Author: It is zero-copy if it gets passed back to the same actor, but will reword.

GPU objects implement the familiar Ray :class:`ObjectRef <ray.ObjectRef>` API but they also:

- Keep data in GPU memory until a transfer is needed
- Avoid expensive serialization and copies to and from CPU memory
- Use efficient data transports like collective communication libraries or RDMA (via `NVIDIA's NIXL <https://github.com/ai-dynamo/nixl>`__) to transfer data between GPUs

.. note::

>    The GPU objects feature is currently in **alpha**. Not all Ray Core APIs are supported yet. Future releases may introduce breaking API changes.

Reviewer: would link to limitations here

Getting started
---------------

> GPU objects currently support ``torch.Tensor`` objects created by Ray actor tasks.

Reviewer: this'll be easy to miss on the page, maybe make it a `.. tip:`

This walkthrough will show how to create and use GPU objects using different *tensor transports*, i.e. the mechanism used to transfer the tensor between actors.

> Currently, we support:

Reviewer: ```suggestion
Currently, GPU objects support:
```

1. `Gloo <https://github.com/pytorch/gloo>`__: A collective communication library for PyTorch and CPUs.
2. `NVIDIA NCCL <https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/index.html>`__: A collective communication library for NVIDIA GPUs.
3. `NVIDIA NIXL <https://github.com/ai-dynamo/nixl>`__ (backed by `UCX <https://github.com/openucx/ucx>`__): A library for accelerating point-to-point transfers via RDMA, especially between various types of memory and NVIDIA GPUs.

For ease of following along, we'll start with the `Gloo <https://github.com/pytorch/gloo>`__ transport, which can be used without any physical GPUs.

Installation
------------

.. note::

>     Under construction.

Reviewer: are these WIP?

Author: Placeholders for future PRs.

.. _gpu-objects-gloo:
Usage with gloo (CPUs only)
---------------------------

Installation
************

.. note::
    Under construction.

Walkthrough
***********

To get started, define an actor class and a task that returns a ``torch.Tensor``:

.. literalinclude:: doc_code/gpu_objects_gloo.py
   :language: python
   :start-after: __normal_example_start__
   :end-before: __normal_example_end__

As written, when the ``torch.Tensor`` is returned, it will be copied into Ray's CPU-based object store.
For CPU-based tensors, this can require an expensive step to copy and serialize the object, while GPU-based tensors additionally require a copy to and from CPU memory.

To enable GPU objects, use the ``tensor_transport`` option in the :func:`@ray.method <ray.method>` decorator.

.. literalinclude:: doc_code/gpu_objects_gloo.py
   :language: python
   :start-after: __gloo_example_start__
   :end-before: __gloo_example_end__

This decorator can be added to any actor tasks that return a ``torch.Tensor``, or that return ``torch.Tensors`` nested inside other Python objects.
Adding this decorator will change Ray's behavior in the following ways:

> 1. When returning the tensor, Ray will store a *reference* to the tensor instead of copying it to CPU memory. If multiple tensors are found, then Ray will store a reference for each tensor.

Reviewer: ```suggestion
1. When returning the tensor, Ray will store a *reference* to the tensor instead of copying it to CPU memory. If multiple tensors are contained in the return object, then Ray will store a reference for each tensor.
```

Reviewer: consider not mentioning multiple tensors here and leaving that to a standalone section

2. When the :class:`ray.ObjectRef` is passed to another task, Ray will use Gloo to transfer the tensor to the destination task.

For (2) to work, we must first create a *collective group* of actors.

Creating a collective group
***************************

.. Add API pages for collective group APIs.

> To create a collective group for use with GPU objects, we must:

Reviewer: ```suggestion
To create a collective group for use with GPU objects:
```

1. Create multiple Ray actors.
2. Create a collective group on the actors using the :func:`ray.experimental.collective.create_collective_group <ray.experimental.collective.create_collective_group>` function. The tensor transport specified should match the one used in the :func:`@ray.method <ray.method>` decorator.

Here is an example:

.. literalinclude:: doc_code/gpu_objects_gloo.py
   :language: python
   :start-after: __gloo_group_start__
   :end-before: __gloo_group_end__

The actors can now communicate directly via gloo.
The group can also be destroyed using the :func:`ray.experimental.collective.destroy_collective_group <ray.experimental.collective.destroy_collective_group>` function.

> After calling this function, the actors can be reused as normal and new collective groups can be created on them.

Reviewer: This makes it seem like actors can't be used for anything else when they're part of a collective group

Passing GPU objects to other actors
***********************************

Now that we have a collective group, we can create and pass GPU objects between the actors.
Here is a full example:

.. literalinclude:: doc_code/gpu_objects_gloo.py
   :language: python
   :start-after: __gloo_full_example_start__
   :end-before: __gloo_full_example_end__

When the :class:`ray.ObjectRef` is passed to another task, Ray will use Gloo to transfer the tensor directly from the source actor to the destination actor instead of the default object store.

> This is done by submitting an additional Ray task to each actor, which executes the send and receive operations on a background thread.

Author: Yes good point, I think it's useful to talk about the internal implementation but it would be better to discuss in its own subsection.

> This is done by submitting an additional Ray task to each actor, which executes the send and receive operations on a background thread.

Reviewer: +1, I'd isolate those bits to an advanced section

> Passing GPU objects to the same actor

Reviewer: ```suggestion
Passing GPU objects to the actor that produced them
```

*************************************

> GPU :class:`ray.ObjectRefs <ray.ObjectRef>` can also be passed to the same actor.

Reviewer: ```suggestion
GPU :class:`ray.ObjectRefs <ray.ObjectRef>` can also be passed to the actor that produced them.
```

> This avoids any copies and just returns the ``torch.Tensors`` directly back to the same actor.

Reviewer: ```suggestion
This avoids any copies and just provides a reference to the same ``torch.Tensor`` that was previously created.
```

For example:

.. literalinclude:: doc_code/gpu_objects_gloo.py
   :language: python
   :start-after: __gloo_intra_actor_start__
   :end-before: __gloo_intra_actor_end__

.. note::
    Ray only keeps a reference to the tensor created by the user, so the tensor objects are *mutable*.
    If ``sender.sum`` were to modify the tensor in the above example, the changes would also be seen by ``receiver.sum``.

>     This differs from the normal Ray Core API, which always makes an immutable copy of data returned by actors.

Reviewer: I feel like this needs to be highlighted more strongly somewhere. Maybe at the top of the page we should have a "TL;DR" that outlines the key points:

- You need to create a collective group and specify a transport
- Objects are mutable until they're consumed
- Otherwise actors can be used as normal

Reviewer: I see, it's outlined in more detail in the limitations section. that is probably sufficient.

still might be nice to have a TL;DR section (and it could link to limitations)

``ray.get``
***********

The :func:`ray.get <ray.get>` function can be used to retrieve the result of a GPU object.

> Note that for collective-based tensor transports (Gloo and NCCL), the :func:`ray.get <ray.get>` function will use the Ray object store to retrieve a copy of the result.

Reviewer: I think this would be better specified in the details of the transport(s)

Author: Hmm this will need to get updated anyway because we are changing the ray.get default to use the one specified in the decorator.

2. The :func:`ray.experimental.collective.create_collective_group <ray.experimental.collective.create_collective_group>` function is used to create a collective group.
3. The tensor is created on the GPU using the ``.cuda()`` method.

Usage with NIXL (CPUs or NVIDIA GPUs)
-------------------------------------

NIXL uses RDMA to transfer data between devices, including CPUs and NVIDIA GPUs.
One advantage of this tensor transport is that it doesn't require a collective group to be created ahead of time.
This means that any actor that has NIXL installed in its environment can be used to create and pass a GPU object.

Here is an example showing how to use NIXL to transfer a GPU object between two actors:

.. literalinclude:: doc_code/gpu_objects_nixl.py
   :language: python
   :start-after: __nixl_full_example_start__
   :end-before: __nixl_full_example_end__

Compared to the :ref:`Gloo example <gpu-objects-gloo>`, the main code differences are:

1. The :func:`@ray.method <ray.method>` uses ``tensor_transport="nixl"`` instead of ``tensor_transport="gloo"``.
2. No collective group is needed.

``ray.get``
***********

> Unlike the collective-based tensor transports (Gloo and NCCL), the :func:`ray.get <ray.get>` function can use NIXL or the Ray object store to retrieve a copy of the result.
> By default, the tensor transport for :func:`ray.get <ray.get>` will be the one specified in the :func:`@ray.method <ray.method>` decorator.

Reviewer: ```suggestion
Unlike the collective-based tensor transports (Gloo and NCCL), the :func:`ray.get <ray.get>` function will use NIXL by default to retrieve a copy of the result.
To override this behavior and use the Ray object store, <...>.
```
